node/server.js

```diff
@@ -8671,57 +8672,46 @@ function init_io() {
 				xy_emit(player, "ui", { type: data.name });
 			} else if (data.name == "3shot" || data.name == "5shot") {
 				player.halt = true;
-				var times = 0;
-				var hit = {};
-				var reftarget = null;
-				var targets = 3;
-				var attack = null;
-				var c_resolve = null;
-				if (data.name == "5shot") {
-					targets = 5;
-				}
+				const targeted = {};
+				let c_resolve = null;
 				//console.log(data.ids);
 				if (is_array(data.ids)) {
-					data.ids.forEach(function (id) {
-						if (times >= targets) {
-							return;
+					// Only look at the first Xshot targets in the array if more are provided
+					for (const id of data.ids.slice(0, data.name === "5shot" ? 5 : 3)) {
+						if (targeted[id]) {
+							// Already targeted this monster
+							continue;
 						}
-						times += 1;
-						var target = instances[player.in].monsters[id];
+						targeted[id] = true;
+						target = instances[player.in].monsters[id];
 						if (!target) {
 							target = instances[player.in].players[id];
 						}
 						if (!target || is_invinc(target) || target.name == player.name) {
-							attack = { failed: true, place: data.name, reason: "no_target" };
-							return;
+							continue;
 						}
-						if (hit[id]) {
-							return;
+						if (isTargetTooFar(target)) {
+							continue;
 						}
-						hit[id] = true;
-						attack = commence_attack(player, target, data.name);
+						const attack = commence_attack(player, target, data.name);
 						if (!attack || !attack.projectile) {
-							return;
+							continue;
 						}
 						if (!c_resolve) {
-							reftarget = target;
+							consume_mp(player, gSkill.mp, target);
```

Comment:
I think this should be moved back out, mp should be consumed even if there are no hits.